protocol/lavaprotocol/finalization_consensus.go

```diff
@@ -100,150 +107,201 @@ func (fc *FinalizationConsensus) insertProviderToConsensus(blockDistanceForFinal
-			if err != nil {
-				// TODO: bring the other data as proof
-				finalizationConflict = &conflicttypes.FinalizationConflict{RelayReply0: reply}
-				// we need to insert into a new consensus group before returning
-				// or create new consensus group if no consensus matched
-				continue
-			}
+	logSuccessUpdate := func() {
+		utils.LavaFormatTrace("finalization information update successfully",
+			utils.LogAttr("specId", fc.specId),
+			utils.LogAttr("finalizationData", finalizedBlocks),
+			utils.LogAttr("currentBlockToHashesToAgreeingProviders", fc.currentEpochBlockToHashesToAgreeingProviders),
+		)
+	}
 
-			if !inserted {
-				// if no discrepency with this group and not inserted yet -> insert into consensus
-				fc.insertProviderToConsensus(blockDistanceForFinalizedData, &consensus, finalizedBlocks, latestBlock, reply, req, providerAddress)
-				inserted = true
-			}
-			// keep comparing with other groups, if there is a new message with a conflict we need to report it too
-		}
-		if !inserted {
-			// means there was a consensus mismatch with everything, so it wasn't inserted and we add it here
-			newHashConsensus := fc.newProviderHashesConsensus(blockDistanceForFinalizedData, providerAddress, latestBlock, finalizedBlocks, reply, req)
-			fc.currentProviderHashesConsensus = append(fc.currentProviderHashesConsensus, newHashConsensus)
-		}
-		if finalizationConflict != nil {
-			// means there was a conflict and we need to report
-			return finalizationConflict, utils.LavaFormatError("Simulation: Conflict found in discrepancyChecker", err)
+	var blockToHashesToAgreeingProviders BlockToHashesToAgreeingProviders
+	foundDiscrepancy, discrepancyBlock := fc.findDiscrepancy(finalizedBlocks, fc.currentEpochBlockToHashesToAgreeingProviders)
+	if foundDiscrepancy {
+		utils.LavaFormatTrace("found discrepancy for provider",
+			utils.LogAttr("specId", fc.specId),
+			utils.LogAttr("currentEpoch", fc.currentEpoch),
+			utils.LogAttr("provider", providerAddress),
+			utils.LogAttr("discrepancyBlock", discrepancyBlock),
+			utils.LogAttr("finalizedBlocks", finalizedBlocks),
+		)
+		blockToHashesToAgreeingProviders = fc.currentEpochBlockToHashesToAgreeingProviders
+	} else {
+		// Could not find discrepancy, let's check with previous epoch
+		foundDiscrepancy, discrepancyBlock = fc.findDiscrepancy(finalizedBlocks, fc.prevEpochBlockToHashesToAgreeingProviders)
+		if foundDiscrepancy {
+			utils.LavaFormatTrace("found discrepancy for provider with previous epoch",
+				utils.LogAttr("specId", fc.specId),
+				utils.LogAttr("currentEpoch", fc.currentEpoch),
+				utils.LogAttr("provider", providerAddress),
+				utils.LogAttr("discrepancyBlock", discrepancyBlock),
+				utils.LogAttr("finalizedBlocks", finalizedBlocks),
+			)
+			blockToHashesToAgreeingProviders = fc.prevEpochBlockToHashesToAgreeingProviders
+		} else {
+			// No discrepancy found, log and return nil
+			logSuccessUpdate()
+			return finalizationConflict, nil
```

Comment:
can we be explicit return nil, nil?

Reply:
Actually, we can't. Linter won't allow it:
<img width="1406" alt="image" src="https://github.com/lavanet/lava/assets/6321801/9aaffdfa-eb72-409a-9ec4-f3e7d9610df3">